cmd/juju/application/bundle.go

```diff
@@ -974,3 +980,145 @@ func processValue(baseDir string, v interface{}) (interface{}, bool, error) {
 
 	return result, true, nil
 }
+
+type bundleConfig struct {
+	Applications map[string]*charm.ApplicationSpec `yaml:"applications"`
+	// TODO soon, add machine mapping and space mapping.
+}
+
+type bundleConfigValueExists struct {
+	Applications map[string]map[string]interface{} `yaml:"applications"`
+}
+
+func processBundleConfig(data *charm.BundleData, bundleConfigFile string) error {
+	if bundleConfigFile == "" {
+		// Nothing to do here.
+		return nil
+	}
+
+	bundleConfigFile, err := utils.NormalizePath(bundleConfigFile)
+	if err != nil {
+		return errors.Annotate(err, "unable to normalise bundle-config file")
+	}
+	// Make sure the filename is absolute.
+	if !filepath.IsAbs(bundleConfigFile) {
+		cwd, err := os.Getwd()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		bundleConfigFile = filepath.Clean(filepath.Join(cwd, bundleConfigFile))
+	}
+	content, err := ioutil.ReadFile(bundleConfigFile)
+	if err != nil {
+		return errors.Annotate(err, "unable to open bundle-config file")
+	}
+	baseDir := filepath.Dir(bundleConfigFile)
+
+	// Now that we have the content, attempt to deserialize into the bundleConfig.
+	var config bundleConfig
+	if err := yaml.Unmarshal(content, &config); err != nil {
+		return errors.Annotate(err, "unable to deserialize config structure")
+	}
+	// If this works, then this deserialisation should certainly succeed.
+	// Since we are only looking to overwrite the values in the underlying bundle
+	// for config values that are set, we need to know if they were actually set,
+	// and not just zero. The configCheck structure is a map that allows us to check
+	// if the fields were actually in the underlying YAML.
+	var configCheck bundleConfigValueExists
+	if err := yaml.Unmarshal(content, &configCheck); err != nil {
+		return errors.Annotate(err, "unable to deserialize config structure")
+	}
+	// Additional checks to make sure that only things that we know about
+	// are passed in as config.
+	var checkTopLevel map[string]interface{}
+	if err := yaml.Unmarshal(content, &checkTopLevel); err != nil {
+		return errors.Annotate(err, "unable to deserialize config structure")
+	}
+	for key := range checkTopLevel {
+		switch key {
+		case "applications":
+			// no-op, all good
+		default:
+			return errors.Errorf("unexpected key %q in config", key)
+		}
+	}
+
+	// We want to confirm that all the applications mentioned in the config
+	// actually exist in the bundle data.
+	for appName, bc := range config.Applications {
+		app, found := data.Applications[appName]
+		if !found {
+			return errors.Errorf("application %q from config not found in bundle", appName)
+		}
+
+		fieldCheck := configCheck.Applications[appName]
+
+		if _, set := fieldCheck["charm"]; set {
```

Comment:
This stuff's kind of messy and a bit gross - it seems like there should be a cleaner way that doesn't require you to know the mapping between attrs and yaml keys and do all the map overlaying, but I can't see what it would be.